### ISSUE TITLE
Optimization using zbuf fails

### ISSUE THREAD

**Issue by TomRoussel** [facebookresearch/pytorch3d#540]
## 🐛 Bugs / Unexpected behaviors

When attempting to optimize a loss that depends on the z-buffer, the loss seems to be maximized instead of minimized.

I discovered this issue while implementing this toy example: given a rendered depthmap (from the z-buffer), find the azimuth that corresponds to this depthmap by iteratively rendering. In other words, we have a goal z-buffer, and we try to use the differentiable renderer to find the azimuth used to render this z-buffer. 

Applying an l2 loss between the goal z-buffer and the current z-buffer causes the optimizer to maximize the loss. Taking the negative of that loss, and optimizing that does result in the correct azimuth being found, which is unexpected.

Is using the z-buffer in such a way not supported or am I doing something wrong?

I implemented the snippet below using the "dolphin.obj" from one of the tutorial notebooks (https://dl.fbaipublicfiles.com/pytorch3d/data/dolphin/dolphin.obj). I installed pytorch3d by compiling it from source, during the week of the 25th of January 2021. 

## Instructions To Reproduce the Issue:

```python
from pytorch3d.io import load_obj
from pytorch3d.structures import Meshes
from pytorch3d.renderer import (
    look_at_view_transform,
    RasterizationSettings,
    MeshRasterizer,
    OpenGLPerspectiveCameras,
)
import torch
import matplotlib.pyplot as plt
from tqdm import tqdm
import numpy as np

device = torch.device("cuda:0")

verts, faces, aux = load_obj("dolphin.obj")

faces_idxs = faces.verts_idx.to(device)
verts = verts.to(device)

car_mesh = Meshes(verts=[verts], faces=[faces_idxs])

elev = 0  # Constant
azim = torch.tensor(101.0, requires_grad=True)  # Optimizable parameter
azim_goal = 130

R_goal, T_goal = look_at_view_transform(
    dist=1, elev=elev, azim=azim_goal, up=torch.Tensor((0, -1, 0)).unsqueeze(0)
)
cameras = OpenGLPerspectiveCameras(device=device, R=R_goal, T=T_goal)

raster_settings_silhouette = RasterizationSettings(
    image_size=256,
    blur_radius=np.log(1.0 / 1e-4 - 1.0) * 1e-4,
    faces_per_pixel=1,
)
rasterizer = MeshRasterizer(
    cameras=cameras, raster_settings=raster_settings_silhouette
)

fragments = rasterizer(car_mesh)
depth_goal = fragments.zbuf[0, :, :, 0].clone().detach()

optim = torch.optim.SGD([azim], lr=1e0)
azims = []
losses = []
for i in tqdm(range(500), desc=f"Azim: {azim}"):
    optim.zero_grad()

    # Update camera extrinsics with the current azimuth angle
    R, T = look_at_view_transform(
        dist=1, elev=elev, azim=azim, up=torch.Tensor((0, -1, 0)).unsqueeze(0)
    )

    fragments = rasterizer(car_mesh, R=R.cuda(), T=T.cuda())

    depth = fragments.zbuf[0, :, :, 0]

    # NOTE: Strange behaviour:
    # Using a negative losses optimizes the azimuth correctly
    # loss = -(depth_goal - depth).pow(2).sum().sqrt()

    # The typical l2 loss however, does not
    loss = (depth_goal - depth).pow(2).sum().sqrt()

    loss.backward()
    optim.step()

    azims.append(float(azim))
    losses.append(loss)

plt.figure()
plt.subplot(121)
plt.plot(losses)
plt.xlabel("Iteration")
plt.ylabel("Loss")

plt.subplot(122)

plt.plot(azims)
plt.xlabel("Iteration")
plt.ylabel("Azimuth")

plt.tight_layout()
plt.show()

```
![Optimizing using L2 loss](https://user-images.githubusercontent.com/21120212/106473205-cdcad100-64a3-11eb-9357-6073368251df.png)

**Comment by nikhilaravi** [comment]
Hi @TomRoussel thanks for writing a very clear description of the issue.  I will try to run your code to repro the issue.

**Comment by nikhilaravi** [comment]
@TomRoussel I think the error in the issue title is misleading as the optimization does not fail but the setup of the problem leads to an error. For the silhouette/depth loss which is what you're using here, we formulate it as `torch.sum((image - image_ref) ** 2)` (and take the mean if it is batched).

Also note that this formulation (azim/elev etc) for optimizing rotations is not particularly stable and is sensitive to initialization position. The tutorial notebooks are only toy examples!

**Comment by TomRoussel** [comment]
@nikhilaravi Thank you for looking into this so quickly! 
I did try several initializations for the azimuth, each time during optimization the loss moved away from target value. I found it extra odd that attempting to optimize the negative L2 loss (which I commented out in the code snippet I posted), resulted in the loss actually converging to the correct orientation! Here's the same graph plotted with that line commented in: (azimuth=130 is the goal)
![Figure_2](https://user-images.githubusercontent.com/21120212/106513138-d4bd0800-64d2-11eb-93f2-767e489b0216.png)

I believe I used the same formulation for the silhouette/depth loss, except I swapped the order of the image and the reference, which shouldn't matter as it is squared anyways.

**Comment by TomRoussel** [comment]
@nikhilaravi After some more experimenting, I have an update. I rewrote the loss and made sure we only consider the error in pixels where the object is in the target depthmap:

```python

diffdepth = ((depth - depth_goal) ** 2)
depthmask = depth_goal != -1
loss = diffdepth[depthmask].mean()
```

Optimizing the loss function converges nicely and returns the correct orientation. 
![Figure_1](https://user-images.githubusercontent.com/21120212/106877327-eb34b080-66d8-11eb-88f9-4b24c8d4150a.png)

Though this does seem to solve the problem, I'm puzzled why the masking is necessary. Do you happen to have some insight why including the -1 depths causes this problem?
